### PR TITLE
Add Avgpool3d and Maxpool3d Layers & Tests

### PR DESCRIPTION
**Pull request by Shashi456**
Before adding tests i wanted to ask, @rxwei can we overload init? Because what we currently support in the max pooling and avg pooling layers is non-square windows, for square windows someone would have to write the layer as 
`MaxPool3d(poolsize:(3,3,3), stride:(2,2,2)`

Is there any way to work around that?

### CODE REVIEW

**Comment by saeta** [comment]
@Shashi456 yes you can overload init, and you can have one init delegate to another init as well. Having convenience initializers to simplify common cases is a good idea.

**Comment by Shashi456** [comment]
Can you explain what you mean by delegate? @saeta

**Comment by saeta** [comment]
Great question! In the Dense layer, we have one initializer call a different initializer to reduce duplication of code: https://github.com/tensorflow/swift-apis/blob/master/Sources/DeepLearning/Layer.swift#L245 Does that make sense?

**Comment by Shashi456** [comment]
Oh that's pretty cool. So you mean in this case, something like: 
public init(poolSize: (Int), strides: (Int), padding: Padding = .valid) {
        self.poolsize = poolsize
        self.stride = stride
        self.padding = padding
    }
{ self.init(poolsize : (poolsize,poolsize,poolsize) ....}

In similar fashion?

**Comment by saeta** [comment]
Yes, and you can even refactor the 2 initializers you currently have in the PR into one delegating (calling) the other. :-)

**Comment by Shashi456** [comment]
You mean initialization delegation is possible across functions?

File: Sources/DeepLearning/Layer.swift [hunk 1]
```diff
@@ -845,6 +845,50 @@ public struct MaxPool2D<Scalar: TensorFlowFloatingPoint>: Layer {
...
+    }
+
+    /// Creates a max pooling layer.
+    ///
+    /// - Parameters:
+    ///   - poolSize: Vertical and horizontal factors by which to downscale.
+    ///   - strides: The strides.
+    ///   - padding: The padding.
+    public init(poolSize: (Int, Int, Int), strides: (Int, Int, Int), padding: Padding = .valid) {
+        self.poolSize = (1, poolSize.0, poolSize.1, poolSize.2, 1)
```
Comment (rxwei) [inline Sources/DeepLearning/Layer.swift]: As @saeta suggested, if you already have a member wise initializer (`(init(poolSize:strides:padding:)`), calling that one via `self.init` is slightly better than initializing every member manually.

File: Sources/DeepLearning/Layer.swift [hunk 2]
```diff
@@ -927,6 +971,50 @@ public struct AvgPool2D<Scalar: TensorFlowFloatingPoint>: Layer {
...
+    /// Returns the output obtained from applying the layer to the given input.
+    ///
+    /// - Parameter input: The input to the layer.
+    /// - Returns: The output.
+    @differentiable
+    public func call(_ input: Tensor<Scalar>) -> Tensor<Scalar> {
+        return input.averagePooled(kernelSize: poolSize, strides: strides, padding: padding)
+    }
+}
+
```
Comment (rxwei) [inline Sources/DeepLearning/Layer.swift]: Remove extra line.

File: Sources/DeepLearning/Layer.swift [hunk 3]
```diff
@@ -845,6 +845,50 @@ public struct MaxPool2D<Scalar: TensorFlowFloatingPoint>: Layer {
...
+    /// The strides of the sliding window for each dimension of a 5-D input.
+    /// Strides in non-spatial dimensions must be `1`.
+    @noDerivative let strides: (Int, Int, Int, Int, Int)
+    /// The padding algorithm for pooling.
+    @noDerivative let padding: Padding
+
+    /// Creates a max pooling layer.
+    public init(
+        poolSize: (Int, Int, Int, Int, Int),
+        strides: (Int, Int, Int, Int, Int),
```
Comment (jekbradbury) [inline Sources/DeepLearning/Layer.swift]: I think I would suggest shifting the stored sizes and strides to be the tuples of three spatial sizes/strides rather than all 5; that way you don't need two `init`s and can avoid exposing the strangeness of the 5-tuple API to the user at all. (I recognize that you've done it this way because that's how our 2D pooling layers work, but those should ideally be changed too!)

Comment (dan-zheng) [reply Sources/DeepLearning/Layer.swift]: @jekbradbury: would it make sense to change operators (e.g. `convolved2D`, `maxPooled`) to not take extra spatial sizes/strides?

```swift
// Current:
func maxPooled(
    kernelSize: (Int, Int, Int, Int),
    strides: (Int, Int, Int, Int),
    padding: Padding
) -> Tensor

// After (implementation uses 1 for first and last size/stride):
func maxPooled(
    kernelSize: (Int, Int),
    strides: (Int, Int),
    padding: Padding
) -> Tensor

// Same for `convolved2D` and `averagePooled`.
```

Comment (rxwei) [reply Sources/DeepLearning/Layer.swift]: If I recall correctly, the fact it took a full tuple was because of historical GPE reasons.

**Comment by Shashi456** [comment]
@rxwei so the things i have to incorporate is :
- Tests for average and max pooling layers 
- Add another init which reference to the original init

would that be it?

**Comment by Shashi456** [comment]
@dan-zheng @saeta @rxwei please review!

File: Sources/DeepLearning/Layer.swift [hunk 4]
```diff
@@ -845,6 +845,62 @@ public struct MaxPool2D<Scalar: TensorFlowFloatingPoint>: Layer {
...
+    }
+
+    /// Creates a max pooling layer.
+    ///
+    /// - Parameters:
+    ///   - poolSize: Vertical and horizontal factors by which to downscale.
+    ///   - strides: The strides.
+    ///   - padding: The padding.
+    public init(poolSize: (Int, Int, Int), strides: (Int, Int, Int), padding: Padding = .valid) {
+        self.poolSize = (1, poolSize.0, poolSize.1, poolSize.2, 1)
```
Comment (dan-zheng) [inline Sources/DeepLearning/Layer.swift]: Please rewrite this initializer using a `self.init` call.
```suggestion
        self.init(poolSize: (1, poolSize.0, poolSize.1, poolSize.2, 1),
                  strides: (1, strides.0, strides.1, strides.2, 1),
                  padding: padding)
```

Please do the same for initializers of `MaxPool2D`, `AvgPool2D`, `MaxPool3D`, and `AvgPool3D`.

File: Sources/DeepLearning/Layer.swift [hunk 5]
```diff
@@ -927,6 +983,62 @@ public struct AvgPool2D<Scalar: TensorFlowFloatingPoint>: Layer {
...
+    @differentiable
+    public func call(_ input: Tensor<Scalar>) -> Tensor<Scalar> {
+        return input.averagePooled(kernelSize: poolSize, strides: strides, padding: padding)
+    }
+}
+
+public extension AvgPool3D {
+  /// Creates a AvgPool3D Layer with the specified poolsize and strides. The poolsize
+  /// and strides are square windows.
+  init(poolSize: Int, stride: Int, padding: Padding = .valid)
```
Comment (dan-zheng) [inline Sources/DeepLearning/Layer.swift]: Please be mindful of the following:
- Use 4-space indentation.
- Wrap `{` onto the previous line.
- Local variable bindings like `let poolsize = poolSize` are not meaningful, use the argument `poolSize` directly.

I hope those points make sense! Being mindful of them helps us merge your PR more quickly.

Please use the following:
```swift
public extension AvgPool3D {
    /// Creates an average pooling layer with the specified pooling window size and stride. The
    /// pooling window is square.
    init(poolSize: Int, stride: Int, padding: Padding = .valid) {
        self.init(poolSize: (poolSize, poolSize, poolSize),
                  strides: (stride, stride, stride),
                  padding: padding)
    }
}
```

File: Tests/DeepLearningTests/LayerTests.swift [hunk 6]
```diff
@@ -34,6 +34,22 @@ final class LayerTests: XCTestCase {
...
+    func testMaxPool2D() {
+        let layer = MaxPool2D<Float>(poolSize:(2, 2), strides:(1, 1), padding:.valid)
+        let input = Tensor(shape: [1, 2, 2, 1], scalars: (0..<4).map(Float.init))
+        let output = layer.inferring(from: input)
+        let expected = Tensor<Float>([[[[3]]]])
+        XCTAssertEqual(round(output), expected)
+    }
+
+    func testMaxPool3D() {
+        let layer = MaxPool3D<Float>(poolSize:(2 ,2, 2), strides:(1, 1, 1), padding:.valid)
```
Comment (dan-zheng) [inline Tests/DeepLearningTests/LayerTests.swift]: Please use correct spacing.
```suggestion
        let layer = MaxPool3D<Float>(poolSize: (2, 2, 2), strides: (1, 1, 1), padding: .valid)
```

Please also fix spacing in other functions (`testMaxPool2D`, `testMaxPool3D`, `testAvgPool2D`).

File: Sources/DeepLearning/Layer.swift (hunk 5 above)
Comment (Shashi456) [reply Sources/DeepLearning/Layer.swift]: Yes, it does. I will pay heed from the next time, Sorry!

Comment (Shashi456) [reply Sources/DeepLearning/Layer.swift]: Also @dan-zheng can you think of any other variations of inits and should i extend an init for 2d max and average pooling as well?

Comment (dan-zheng) [reply Sources/DeepLearning/Layer.swift]: Note that this API is not comprehensive: both pooling window sizes and strides can either be specified as "multiple integers" or "one integer, implicitly duplicated".

In other words, all of the following may be useful:
- `init(poolSize: (Int, Int, Int), strides: (Int, Int, Int), padding: Padding)`
- `init(poolSize: Int, strides: Int, padding: Padding)`
- `init(poolSize: (Int, Int, Int), strides: Int, padding: Padding)`
- `init(poolSize: Int, strides: (Int, Int, Int) padding: Padding)`

Something to think about and visit after this PR.
Here's Keras (note "integer or tuple of two integers"):

image

Comment (Shashi456) [reply Sources/DeepLearning/Layer.swift]: While writing the code for this, i thought about exactly this case, but had no idea how to address it exactly. In python, It is easier to handle. I'm not very well versed with init overloading in swift

Comment (dan-zheng) [reply Sources/DeepLearning/Layer.swift]: > Yes, it does. I will pay heed from the next time, Sorry!

Thanks a ton! That would be a huge help. 🙂

> Also @dan-zheng can you think of any other variations of inits and should i extend an init for 2d max and average pooling as well?

I just commented above about this.
To match Keras, we could have the following combinations:
- `poolSize`: `(Int, Int, Int)` or `Int`
- `strides`: `(Int, Int, Int)` or `Int?` (if `nil`, then use `poolSize` as strides)

**Comment by Shashi456** [comment]
@dan-zheng thank you for being so patient with these reviews.

**Comment by Shashi456** [comment]
@rxwei can you review this once more?

**Comment by rxwei** [comment]
This should not have been merged because it does not compile. The `self.init` statement cannot appear on top level. The code does not compile.

@Shashi456 Could you make sure everything compiles with the latest toolchain before you commit?

**Comment by rxwei** [comment]
I reverted it in #118. Feel free to start a new PR when you have a fix.

**Comment by Shashi456** [comment]
@rxwei when i just try to simply add a layer in colab, to get an idea of how the layer works. It often throws me this error, could you tell me how i can solve this ?

```
error: <Cell 8>:2:15: error: type 'MaxPool3D<Scalar>' does not conform to protocol 'Layer'
public struct MaxPool3D<Scalar: TensorFlowFloatingPoint>: Layer {
              ^

TensorFlow.Layer:2:20: note: protocol requires nested type 'Input'; do you want to add it?
    associatedtype Input : Differentiable
                   ^

TensorFlow.Layer:3:20: note: protocol requires nested type 'Output'; do you want to add it?
    associatedtype Output : Differentiable
```

**Comment by Shashi456** [comment]
Is there a temporary fix I could do @rxwei?

**Comment by rxwei** [comment]
Could you share a code example that I use to reproduce your error?

**Comment by Shashi456** [comment]
@rxwei for example, if i use this code, I know this could also be because max is not differentiable, but this input output error is common with any layer i try to define. 
```
@_fixed_layout
public struct GlobalMaxPooling3D<Scalar: TensorFlowFloatingPoint>: Layer {
    /// Creates a global max pooling layer.
    public init() {}

    /// Returns the output obtained from applying the layer to the given input.
    ///
    /// - Parameters:
    ///   - input: The input to the layer.
    ///   - context: The contextual information for the layer application, e.g. the current learning
    ///     phase.
    /// - Returns: The output.
    @differentiable
    public func applied(to input: Tensor<Scalar>, in _: Context) -> Tensor<Scalar> {
        return input.max(squeezingAxes: [1, 2, 3])
    }
}

```
image

**Comment by rxwei** [comment]
Colab is not yet updated with the latest APIs (v0.3). You can follow the instructions [here](https://github.com/tensorflow/swift/blob/master/Usage.md#getting-newer-swift-builds) to install a new build to Colab.

**Comment by Shashi456** [comment]
Thanks a lot ! I will try and see if this works.

**Comment by rxwei** [comment]
Actually, it should be updated now if you restart Colab.

**Comment by Shashi456** [comment]
@rxwei the error majorly seems to be :
```
error: cannot convert value of type '(Int, Int, Int, Int, Int)' to expected argument type '(Int, Int, Int, Int)'
        return input.maxPooled(kernelSize: poolSize, strides: strides, padding: padding)
```
